Add new recipe for emacs-workspaces

mclearc opened the pull request:
### Brief summary of what the package does

This package leverages tab-bar and project (both built into emacs 27+) to
create buffer-isolated workspaces that also integrate with your
version-controlled projects. It should work with emacs 27+. It is tested to work
with a single frame workflow, but should work with multiple frames as well. 

### Direct link to the package repository

https://github.com/mclear-tools/emacs-workspaces

### Your association with the package

I'm the maintainer

### Relevant communications with the upstream package maintainer

 **None needed** 

### Checklist

<!-- Please confirm with `x`: -->

- [X ] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X ] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X ] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X ] My elisp byte-compiles cleanly
- [X ] `M-x checkdoc` is happy with my docstrings
- [ X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

AkibAzmain commented:
I think the `emacs-` prefix should be removed, because as Emacs Lisp runs on Emacs so its clear that the package is obviously Emacs related.  I would suggest also using `-` instead of `/` in symbol names, as symbol name don't use `/` except exceptional cases (there are 45048 symbols defined in my Emacs at the time of writing this, of which only 340 (about 0.75%) symbols contain `/` in their names).

AkibAzmain commented:
From [Emacs Lisp manual](https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html):

> * Simply loading a package should not change Emacs’s editing
>   behavior.  Include a command or commands to enable and disable the
>   feature, or to invoke it.
>
>   This convention is mandatory for any file that includes custom
>   definitions.  If fixing such a file to follow this convention
>   requires an incompatible change, go ahead and make the incompatible
>   change; don’t postpone it.
>
> * You should choose a short word to distinguish your program from
>   other Lisp programs.  The names of all global symbols in your
>   program, that is the names of variables, constants, and functions,
>   should begin with that chosen prefix.  Separate the prefix from the
>   rest of the name with a hyphen, ‘-’.  This practice helps avoid
>   name conflicts, since all global variables in Emacs Lisp share the
>   same name space, and all functions share another name space(1).
>   Use two hyphens to separate prefix and name if the symbol is not
>   meant to be used by other packages.
>
>   Occasionally, for a command name intended for users to use, it is
>   more convenient if some words come before the package’s name
>   prefix.  For example, it is our convention to have commands that
>   list objects named as ‘list-SOMETHING’, e.g., a package called
>   ‘frob’ could have a command ‘list-frobs’, when its other global
>   symbols begin with ‘frob-’.  Also, constructs that define
>   functions, variables, etc., work better if they start with ‘defun’
>   or ‘defvar’, so put the name prefix later on in the name.
>
>   This recommendation applies even to names for traditional Lisp
>   primitives that are not primitives in Emacs Lisp—such as
>   ‘copy-list’.  Believe it or not, there is more than one plausible
>   way to define ‘copy-list’.  Play it safe; append your name prefix
>   to produce a name like ‘foo-copy-list’ or ‘mylib-copy-list’
>   instead.
>
>   If you write a function that you think ought to be added to Emacs
>   under a certain name, such as ‘twiddle-files’, don’t call it by
>   that name in your program.  Call it ‘mylib-twiddle-files’ in your
>   program, and send mail to ‘bug-gnu-emacs@gnu.org’ suggesting we add
>   it to Emacs.  If and when we do, we can change the name easily
>   enough.
>
>   If one prefix is insufficient, your package can use two or three
>   alternative common prefixes, so long as they make sense.

mclearc commented:
I've addressed both of these in https://github.com/mclear-tools/emacs-workspaces/commit/91d2fdb8e94f5431b49669a829b4b7cd83c2c589. I use "/" for personal interactive functions in my config and had just thought to carry it over here. But happy to make the changes.

AkibAzmain commented:
But you still use [advice function during load](https://github.com/mclear-tools/emacs-workspaces/blob/27c677336c329e8f1b7fca1b7480af92df47615c/workspaces.el#L119), you should use a global minor mode.

AkibAzmain commented:
But you still [advice function during load](https://github.com/mclear-tools/emacs-workspaces/blob/27c677336c329e8f1b7fca1b7480af92df47615c/workspaces.el#L119), you should use a global minor mode.

mclearc commented:
Ok -- we should hopefully be good to go now. See https://github.com/mclear-tools/emacs-workspaces/commit/2be2233ebe947d6c34e9181eca916b01536c997a for all the relevant changes. I created a minor-mode, got rid of advice, added relevant new functions and var, and added a keymap.

riscy commented:
Thanks @AkibAzmain -- great advice. :)

Here's a very quick first pass from my end, @mclearc --

`workspaces.el` with byte-compile using Emacs 27.1:
```
workspaces.el:163:1:Warning: Unused lexical variable
    `project-current-inhibit-prompt'
In end of data:
workspaces.el:301:1:Warning: the following functions are not known to be
    defined: project-prompt-project-dir, project-remember-project,
    project-vc-dir, dired-jump-other-window
```
Probably you need a `(require 'project)` at the top-level.

`workspaces.el` with checkdoc 0.6.2:
```
workspaces.el:145: First line is not a complete sentence
workspaces.el:180: First line is not a complete sentence
workspaces.el:202: First line is not a complete sentence
workspaces.el:229: First line is not a complete sentence
```

`workspaces.el` with [melpazoid](https://github.com/riscy/melpazoid):
- workspaces.el#L176: It's safer to sharp-quote function names; use `#'`
- workspaces.el#L192: It's safer to sharp-quote function names; use `#'`
- workspaces.el#L222: It's safer to sharp-quote function names; use `#'`
- workspaces.el#L252: It's safer to sharp-quote function names; use `#'`

Package and license:
- No .el file matches the name 'emacs-workspaces' (probably the recipe needs updating)

mclearc commented:
Thank you @riscy. I think commit https://github.com/mclear-tools/emacs-workspaces/commit/4a4f623299df656e6cbaec046635459b45f26614 addresses everything, though I haven't been able to reproduce any byte-compile warnings about project.el on emacs 27, 28, or master. I already had `(require 'project)` at the top-level, so not sure what is going on there?

riscy commented:
My mistake - I think the version of `project` you must have installed is later than the 27.1 version; there's a newer version hosted on ELPA that you could declare a dependency on.  For `dired-jump-other-window` you'd want `(require 'dired-x)`

For some reason my script didn't pick these up earlier today but there is also a similarly named package, and it might be worthwhile to avoid a namespace collision:

- workspaces: https://github.com/emacsmirror/emacswiki.org/blob/master/workspaces.el

mclearc commented:
> My mistake - I think the version of project you must have installed is later than the 27.1 version; there's a newer version hosted on ELPA that you could declare a dependency on. For dired-jump-other-window you'd want (require 'dired-x)

I haven't checked emacs-27 but isn't this in dired and not dired-x? It is for emacs on master. I'll update the version of project.el to "0.8.1", which I think is the most recent one. 

>For some reason my script didn't pick these up earlier today but there is also a similarly named package, and it might be >worthwhile to avoid a namespace collision:
>
>workspaces: https://github.com/emacsmirror/emacswiki.org/blob/master/workspaces.el

I'm not quite sure what to do about this. This was partly why I had the name "emacs-workspaces". Is the suggestion that I change the namespace further? Do you have any suggestions? Thanks.

mclearc commented:
Ok -- added dired-x as explicit requirement and a more recent version of project.el. Also have a new namespace that should not collide with any pre-existing package. I've revised the recipe as well. See commit https://github.com/mclear-tools/tabspaces/commit/910898dcdbff646efdd75bd87fe316dfa9987b4f for details.

riscy commented:
Thanks for accommodating, especially with the naming of the package.  Everything looks great here -- happy to merge.  And welcome to MELPA. :)

I'll leave a couple more minor comments here that are non-blocking:
- The use of `progn` in some functions is unnecessary I think (e.g. `tabspaces-open-existing-project-and-workspace`, `tabspaces-create-new-project-and-workspace`, the `let` in `tabspaces--create-new-vc-project`)
- I think this form `(if (featurep 'magit) (progn (require 'magit) (magit-status-setup-buffer)) ...)` won't work as you intend; `(featurep 'magit)` will only return non-nil if `magit` has already been required, so at least the `require` isn't necessary
